exercises/06-redux-saga/src/modules/users/users-saga.ts

```diff
@@ -0,0 +1,42 @@
+import { all, call, fork, put, takeEvery } from 'redux-saga/effects'
+import { Action } from 'redux'
+
+import { usersActions } from 'modules/users/users-slice'
+import * as UsersEffects from 'modules/users/users-effects'
+
+function* getUsers() {
+  try {
+    const { data: users } = yield call(UsersEffects.getUsers)
+    yield put(usersActions.usersLoaded(users))
+  } catch (error) {
+    console.log(error.message)
+  }
+}
+
+function* addUser(action: Action) {
```

Comment:
You can use `function* addUser(action: ReturnType<typeof usersActions.addUser>) {`
And I do not think that check below is needed? :wonder:

Reply:
I can't seem to find any other way for sagas to make good checks for Action types :(   

With the `function* addUser(action: ReturnType<typeof usersActions.addUser>) {` approach:
- since the type is given, TypeScript expects an action to contain specific properties - we are able to call `action.payload` in the generator body.
- but no checks are done at the point of assigning saga to the action: I could easily call `yield all([takeEvery(usersActions.usersLoaded.type, addUser)])`, which assigns `addUser` saga to the incorect Action type. This will get zero static warnings and errors, the code just won't work at runtime: either silently (when an action does have the payload with expected properties) or with errors which occur later on. 
- the Action types produced by Redux Toolkit are not literally typed (it's possible with `createAction`, but not with `createSlice`), so the `ReturnType` would not check that the action is in fact of correct type, only that it has the correct properties 

On the other hand the `.match` [approach](https://redux-toolkit.js.org/usage/usage-with-typescript#alternative-to-using-a-literally-typed-actiontype):
- will check that the actual Action type match the type of passed `action`
- will allow access to correct action properties
- will allow to at least log an error at runtime which pinpoints to the source of the problem


I figured, that the `.match()` approach would be safer and went with that. If there's a better approach I'll be happy to learn about it.

Reply:
Aha, it worked for me because I was using typesafe-actions :/
Shame, if there is no better way for Redux Toolkit.
I will try to check...

Reply:
Hmm, can't find any other solution for saga, another point for using observables instead of sagas - https://redux-toolkit.js.org/api/createAction#with-redux-observable :/

But maybe having a helper function like
```
const createSagaHandler = <P, T extends string, A, B, C>(actionCreator: ActionCreatorWithPayload<P, T>, perform: (a: PayloadAction<P, T>) => Generator<A,B,C>) => function* addUser(...params: Parameters<typeof perform>) {
  if (!actionCreator.match(params[0])) {
    console.error(
      'Unexpected type',
      // @ts-ignore
      `'${params[0].type}'`,
      'was passed to addUser saga.'
    )
    return
  }
  yield *perform(...params)
}

const addUser = createSagaHandler(usersActions.addUser, function* ({ payload: user }) {
  try {
    const response = yield call(UsersEffects.addUser, user)

    if (response) {
      yield fork(getUsers)
    }
  } catch (error) {
    console.log(error.message)
  }
})
```

Reply:
I was a bit disappointed in this `Redux Toolkit + Redux-Saga` combination to be honest.

This solution looks good for me (if it means anything :smile:), but it may be too complex for training exercises. What do you think?

Reply:
You are right, for first access it would be better to not overwhelm the reader with such a complexly typed function.
But you can try to sneak it in further exercises :)

Reply:
I discussed this briefly with @tomkis as well. We agree that Sagas are not great with TypeScript 😄  But there is probably nothing that we can do about that.

Reply:
@branoholy I'm not sure. Even if Redux Toolkit types were better or the other tool would have been used, the sagas typing won't check `takeEvery` consistency, will it?

Redux Toolkit will at least allow a runtime check

Reply:
@karelklima I think the only problem with Sagas and TS is that the `yield`ed values lose types (as I mentioned in https://github.com/salsita/react-training/pull/55#discussion_r469983424). Sagas can do nothing with this. This is a "problem" with TS itself. The rest should be fine I guess. :thinking: The typing is actually pretty good. At least I don't remember any other problem but my memory isn't good. :smiley: 

@dinassh I have only experience with https://github.com/aikoven/typescript-fsa but their typing is good enough even for `takeEvery`. See the following code and the type errors below. Am I missing something?

```ts
import { takeEvery } from "redux-saga/effects";
import actionCreatorFactory, { Action } from "typescript-fsa";

const actionCreator = actionCreatorFactory();

interface FooPayload {
  foo: string;
}
const createFoo = actionCreator<FooPayload>("FOO");

interface BarPayload {
  bar: number;
}
const createBar = actionCreator<BarPayload>("BAR");

const createBaz = actionCreator("BAZ");

// ------------------------------------------------

function* fooSaga(action: Action<FooPayload>) {
  console.log(action.type);
}

function* barSaga(action: Action<BarPayload>) {
  console.log(action.type);
}

function* bazSaga() {}

takeEvery(createFoo, fooSaga);
takeEvery(createFoo, barSaga);
takeEvery(createFoo, bazSaga);

takeEvery(createBar, fooSaga);
takeEvery(createBar, barSaga);
takeEvery(createBar, bazSaga);

takeEvery(createBaz, fooSaga);
takeEvery(createBaz, barSaga);
takeEvery(createBaz, bazSaga);

```

![image](https://user-images.githubusercontent.com/4808634/90251422-fef40e00-de3d-11ea-9b3e-fe66f6869d4f.png)

Reply:
Thanks for the example! It wasn't working for me and led me to the true source of the problem. The problem was with an old version of `redux-saga` package (it was not updated yet due to compatibility with `react-modules`). I'm so very sorry for not noticing it earlier!

The same behavior may be achieved with Redux Toolkit after all:
```ts
import { createAction, PayloadAction } from '@reduxjs/toolkit'

interface FooPayload {
  foo: string
}
const createFoo = createAction<FooPayload>('FOO')

interface BarPayload {
  bar: number
}
const createBar = createAction<BarPayload>('BAR')

const createBaz = createAction<void>('BAZ')

// ------------------------------------------------

function* fooSaga(action: PayloadAction<FooPayload>) {
  console.log('foo', action.type)
  yield
}

function* barSaga(action: PayloadAction<BarPayload>) {
  console.log('bar', action.type)
  yield
}

function* bazSaga() {
  yield
}

export function* fooBazBarSaga() {
  yield all([takeEvery(createFoo, fooSaga)])
  yield all([takeEvery(createFoo, barSaga)])
  yield all([takeEvery(createFoo, bazSaga)])

  yield all([takeEvery(createBar, fooSaga)])
  yield all([takeEvery(createBar, barSaga)])
  yield all([takeEvery(createBar, bazSaga)])

  yield all([takeEvery(createBaz, fooSaga)])
  yield all([takeEvery(createBaz, barSaga)])
  yield all([takeEvery(createBaz, bazSaga)])
}
```

This piece of code results in same errors, as your example.

It can be improved if used with Redux Toolkit strongly typed `createAction` (it won't work with `createSlice`):
```ts
import { takeEvery } from 'redux-saga/effects'
import { createAction } from '@reduxjs/toolkit'

interface FooPayload {
  foo: string
}
const createFoo = createAction<FooPayload, 'FOO'>('FOO')
const createBar = createAction<FooPayload, 'BAR'>('BAR')
const createBaz = createAction<void, 'BAZ'>('BAZ')

// ------------------------------------------------

function* fooSaga(action: ReturnType<typeof createFoo>) {
  console.log('foo', action.type)
  yield
}

function* barSaga(action: ReturnType<typeof createBar>) {
  console.log('bar', action.type)
  yield
}

function* bazSaga(action: ReturnType<typeof createBaz>) {
  yield
}

export function* fooBazBarSaga() {
  yield all([takeEvery(createFoo, fooSaga)])
  yield all([takeEvery(createFoo, barSaga)])
  yield all([takeEvery(createFoo, bazSaga)])

  yield all([takeEvery(createBar, fooSaga)])
  yield all([takeEvery(createBar, barSaga)])
  yield all([takeEvery(createBar, bazSaga)])

  yield all([takeEvery(createBaz, fooSaga)])
  yield all([takeEvery(createBaz, barSaga)])
  yield all([takeEvery(createBaz, bazSaga)])
}
```
which results in
![image](https://user-images.githubusercontent.com/67343955/90331233-2cae9380-dfb3-11ea-93fa-e80a3bc0ab56.png)

Reply:
Fixed in code: 2486f14f0c6055221376a8f1e7fe056c2c4d39dd and in tutorial: ad0a6feb639ef987a3f7642548f4a6839a5dfff1

Fixed in slides, see [link to affected page](https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs/edit#slide=id.g364d1c8966_0_34)

Reply:
Great! :+1: